include/SFML/Config.hpp

```diff
@@ -43,8 +43,22 @@
 
     // Windows
     #define SFML_SYSTEM_WINDOWS
+
     #ifndef NOMINMAX
-        #define NOMINMAX
+        #error 'NOMINMAX' must be defined to avoid issues with 'windows.h'
+    #endif
+
+    #ifndef WIN32_LEAN_AND_MEAN
+        #error 'WIN32_LEAN_AND_MEAN' must be defined to limit 'windows.h' compile-time overhead
+    #endif
+
+    #if !defined(_WIN32_WINDOWS) || !defined(_WIN32_WINNT) || !defined(WINVER) || \
+        (_WIN32_WINDOWS != 0x0501) || (_WIN32_WINNT != 0x0501) || (WINVER != 0x0501)
+        #error Both '_WIN32_WINDOWS' and '_WIN32_WINNT' must be defined '0x0501'
+    #endif
+
+    #if !defined(DIRECTINPUT_VERSION) || (DIRECTINPUT_VERSION != 0x0800)
+        #error 'DIRECTINPUT_VERSION' must be defined as '0x0800'
     #endif
```

Comment:
Is this really required? This file is part of the public interface and is directly/indirectly included by any file as soon as the user touches an SFML header in their code. All these definitions just affect which parts of the Windows headers are active. The nice thing about the Windows API is that it is a C API which means that we don't have to be too careful about mixing different API versions between SFML and user code. Microsoft has made sure that any additions are backwards compatible.

If you put yourself into the shoes of the user, I would find these compiler errors confusing/annoying, especially for a beginner who had no intention of doing Windows programming in their own code (that's what they chose SFML for). Just including an SFML header will almost always generate these 4 errors for them if they don't explicitly add these definitions to their project.

If the goal was to ensure that we build SFML with the same definitions everywhere then this has to be done in an internal header that is not exposed via the public interface.

Reply:
I understand your concern, and I believe you're right.

> If the goal was to ensure that we build SFML with the same definitions everywhere then this has to be done in an internal header that is not exposed via the public interface.

That was indeed the goal. How I should I call that internal header and where should I put it? Can I create a `Priv` subfolder under `SFML`?

Reply:
Maybe just dump it into src/SFML? 😁

When it comes to such things, I am not the right guy to ask.

@SFML/sfml

Reply:
I would actually like to have a `Priv` subfolder for implementation-only headers. It would be extremely useful when trying to reuse small utilities between different SFML components. I am actually surprised there is nothing like that, and I've seen some repetition of internal utilities due to the lack of such structure.

How do you feel about `SFML/Priv/WindowsHeader.h`? And then we could include that instead of `<windows.h>` whenever required.

Reply:
`<windows.h>` is a special unicorn. Unlike the Posix headers where you can include what you use (and nothing more) the general expectation when developing for Windows is that you just `#include <windows.h>` and that is it. This is also the reason why they have tried (e.g. WIN32_LEAN_AND_MEAN) to retroactively fix the problem they created, a problem that does not exist on other platforms. I am just a bit concerned that the contents of this directory would be very Windows heavy and there would not be analogous files for other platforms.

Reply:
@binary1248: let's discuss on Slack please.